streamer/src/nonblocking/qos.rs

```diff
@@ -0,0 +1,64 @@
+use {
+    crate::nonblocking::{
+        quic::{ClientConnectionTracker, ConnectionPeerType},
+        stream_throttle::ConnectionStreamCounter,
+    },
+    quinn::Connection,
+    std::{sync::Arc, time::Duration},
+    tokio_util::sync::CancellationToken,
+};
+
+/// A trait to provide context about a connection, such as peer type,
+/// remote pubkey. This is opaque to the framework and is provided by
+/// the concrete implementation of QosController.
+pub(crate) trait ConnectionContext: Clone + Send + Sync {
+    fn peer_type(&self) -> ConnectionPeerType;
+    fn remote_pubkey(&self) -> Option<solana_pubkey::Pubkey>;
+}
+
+/// A trait to manage QoS for connections. This includes
+/// 1) deriving the ConnectionContext for a connection
+/// 2) managing connection caching and connection limits, stream limits
+pub(crate) trait QosController<C: ConnectionContext> {
+    /// Derive the ConnectionContext for a connection
+    fn derive_connection_context(&self, connection: &Connection) -> C;
+
+    /// Try to add a new connection to cache. If successful, return a CancellationToken and
+    /// a ConnectionStreamCounter to track the streams created on this connection.
+    /// Otherwise return None.
+    fn try_cache_connection(
```

Comment:
Technically we are not caching connections, we are adding them to a connection table (or not). maybe just call it try_add_connection?

Also comment does not match the signature (it returns a future not an Option)